api/api.go

```diff
@@ -114,3 +126,40 @@ func NewAPI(config *conf.Configuration, db *gorm.DB, mailer *mailer.Mailer) *API
 	api.handler = corsHandler.Handler(mux)
 	return api
 }
+
+func (a *API) trace(f func(context.Context, http.ResponseWriter, *http.Request)) func(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) {
+		id := uuid.NewRandom().String()
+		log := a.log.WithField("request_id", id)
+
+		log = log.WithFields(logrus.Fields{
+			"method": r.Method,
+			"path":   r.URL.Path,
+		})
+
+		// optionally add the user id stuff
+		claims := getClaims(ctx)
+		if claims != nil {
+			log = log.WithFields(logrus.Fields{
+				"claim_id":     claims.ID,
+				"claim_groups": claims.Groups,
+				"claim_email":  claims.Email,
+			})
+		}
+
+		ctx = withRequestID(ctx, id)
+		ctx = withLogger(ctx, log)
+		ctx = withConfig(ctx, a.config)
+
+		rp := ResponseProxy{ResponseWriter: w}
+		startTime := time.Now()
+
+		log.Info("request started")
+		defer log.WithFields(logrus.Fields{
+			"status":   rp.statusCode,
```

Comment:
I'm pretty sure these variables are evaluated when the defer is defined and not when it's executed, so statusCode and duration will always be 0:

https://play.golang.org/p/lhee3Pqc8G